gh-85076: Document exceptions that can be raised by importlib.import_module (#94662)
python/cpython: Doc/library/importlib.rst

    need to call :func:`invalidate_caches` in order for the new module to be
    noticed by the import system.

>     If the module cannot be imported, :func:`import_module` will raise an
>     :exc:`ImportError`. If the module is found and loaded, but the code in the
>     module raises an exception, :func:`import_module` will pass that exception
>     to the module that called it.

Reviewer: ```suggestion
    If the module cannot be found or loaded, :func:`import_module` will raise an
    :exc:`ImportError`.
```

"cannot be imported" is very generic and could conceivably apply to explicitly raised exceptions, like in the second half of this paragraph, so makes sense to be more specific.

I think we could just omit the second half of the paragraph. Talking about "passing exceptions" makes it sound more complicated than it is. The behaviour here is intuitive and already documented by the sentence "This means all semantics of the function are derived from :func:`importlib.__import__`."

Author: Doesn't "cannot be found or *loaded*" imply that that's the behavior if the module cannot be loaded because code in the module raised an exception? Maybe "If the module cannot be found or its files cannot be accessed..." instead? That makes it clear that ImportError is raised in cases like the module not being found or being inaccessible due to file permissions, not cases where it is found but not successfully loaded.

Reviewer 2: I suggest to look further down, where I tried to consolidate things. (sorry for the spam, I hope it's helpful for readers 😅)

>     If the module cannot be imported, :func:`import_module` will raise an
>     :exc:`ImportError`. If the module is found and loaded, but the code in the
>     module raises an exception, :func:`import_module` will pass that exception
>     to the module that called it.

Reviewer 3: ```suggestion
    If the module cannot be imported, :func:`import_module` will raise
    :exc:`ImportError` or an appropriate subclass like
    :exc:`ModuleNotFoundError`.
```

I don't think there's any benefit to saying exceptions will propagate as that's the default for any Python code.

I also think we should make sure that people are aware that subclasses may be raised as well since typically people really want `ModuleNotFoundError`.

Reviewer 2: I suggest to look further down, where I tried to consolidate things. (sorry for the spam, I hope it's helpful for readers 😅)

>     If the module cannot be imported, :func:`import_module` will raise an
>     :exc:`ImportError`. If the module is found and loaded, but the code in the
>     module raises an exception, :func:`import_module` will pass that exception
>     to the module that called it.

Reviewer 2: Trying to incorporate everything discussed at the same time.

@brettcannon's  suggestion:
- is more specific on the type of exception that can be expected.
- does not incorporate the remark from @hauntsaninja, that criticises the fact that "cannot be imported" is very generic.

In the first thread @kj7rrv suggested a nice way of explaining how a found module can't be imported. Not sure if it's now too specific, but let's continue the discussion:

```suggestion
    If the module cannot be found or its files cannot be accessed,
    :func:`import_module` will raise :exc:`ImportError` or an appropriate
    subclass like :exc:`ModuleNotFoundError`.
```
